--- a/source/Cleaner.cpp
+++ b/source/Cleaner.cpp
@@ -1496,7 +1496,10 @@ void Cleaner::removeDuplicates() {
         {
             if (alig->sequences[i] == alig->sequences[x])
             {
-                alig->saveSequences[i] = -1;
+                if (alig->saveSequences[i] != -1) {
+                    alig->saveSequences[i] = -1;
+                    alig->numberOfSequences--;
+                }
                 debug.report(InfoCode::RemovingDuplicateSequences,
                         new std::string[2] {
                     alig->seqsName[i],
